e2e: node: fix the device plugin test

ffromani opened the pull request:
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:
The DevicePluginProbe test was constantly failing lately.
One of the key reasons was that the testing pod was restarting too often.
In general this test suffered a bit of bit rot, but we postpone larger cleanups until the kubelet-node-serial lane is reliable again.

#### Which issue(s) this PR fixes:
https://github.com/kubernetes/kubernetes/issues/103691

#### Special notes for your reviewer:
**PARTIALLY** Fixes https://github.com/kubernetes/kubernetes/issues/102148 - Other failure in the serial lane will be addressed by other PRs.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial

fromanirh commented:
/sig node

fromanirh commented:
/hold
I can reproduce locally and this patch fixes (waiting for the lanes to confirm) but I need to re-upload to provide more context about
- why the test was failing
- why this fix fixes - and how it could and should be improved

fromanirh commented:
it seems to be working, the serial lane is now failing for
```
I0629 14:09:35.261] [sig-node] CriticalPod [Serial] [Disruptive] [NodeFeature:CriticalPod]
I0629 14:09:35.261] _output/local/go/src/k8s.io/kubernetes/test/e2e_node/framework.go:23
I0629 14:09:35.261]   when we need to admit a critical pod
I0629 14:09:35.261]   _output/local/go/src/k8s.io/kubernetes/test/e2e_node/critical_pod_test.go:46
I0629 14:09:35.261]     should be able to create and delete a critical pod [It] 
```
which will be handled on a separate PR

fromanirh commented:
/hold cancel
/test pull-kubernetes-node-kubelet-serial

ehashman commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -112,13 +110,6 @@ func readDaemonSetV1OrDie(objBytes []byte) *appsv1.DaemonSet {
-			if initialConfig.FeatureGates == nil {
-				initialConfig.FeatureGates = map[string]bool{}
-			}
-			initialConfig.FeatureGates[string(features.KubeletPodResources)] = true
```
Was this feature flag removed? We don't seem to be replacing this with anything.

fromanirh replied:
yes, better to move it back until at least the cpumanager feature gate is locked to true

ehashman commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -152,11 +143,16 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+			// amount of time so we have restarts - which we check later in the flow.
+			// A probably better approach would be to trigger restart from the outside
+			// (ExecCommandOnContainer?)
+			podRECMD := "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo stub devices: $devs && sleep 1m"
```
I think it'd probably be better to call the wait from the go code rather than shelling out, I almost completely missed that we added a sleep here.

fromanirh replied:
is going to be challenging because this is the entry point of the test container (see makeBusyboxPod), and we need to work with the images we have available.

ehashman reviewed (commented):
/priority important-soon
/triage accepted

ffromani committed:
e2e: node: fix the device plugin test

The DevicePluginProbe test was constantly failing lately.
One of the key reasons was that the testing pod was restarting
too often, too fast to be reported as Ready, which is one of the
first thing it is checked in the test.

What we need here is
- run a test pod, which must wait long enough for the pod to
  be reported Ready, BUT
- the test pod should wait for a fixed amount of time so we have
   restarts - we need them to test the device allocation.
- (not strictly needed to fix the broken test, but helpful anyway)
  add the sample device plugin to the list of pre-pulled images.

In general this test suffered a bit of bit rot, but
we postpone larger cleanups until the kubelet-node-serial lane
is reliable again.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial

ehashman commented:
/assign

ehashman commented:
/assign @swatisehgal

ehashman commented:
/test pull-kubernetes-node-kubelet-serial

ehashman reviewed (commented):
If this fixes the test, I am okay with this :+1: 

Waiting on the latest serial run, it wasn't clear what the failures were in the last run, but I think it was just the criticalpod tests.

ehashman commented:
Can we edit the description to say this fixes https://github.com/kubernetes/kubernetes/issues/103691 ?

ehashman commented:
Looks like this is now failing for a new, exciting reason? https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103297/pull-kubernetes-node-kubelet-serial/1415379089107193856/#1:build-log.txt%3A1887

```
W0714 19:36:36.926]     Jul 14 19:29:12.189: Test file "test/e2e/testing-manifests/sample-device-plugin.yaml" was not found.
      
W0714 19:36:36.927]     The following files are embedded into the test executable:
W0714 19:36:36.927]     	test/e2e_node/testing-manifests/sriovdp-cm.yaml
W0714 19:36:36.927]     	test/e2e_node/testing-manifests/sriovdp-ds.yaml
W0714 19:36:36.927]     	test/e2e_node/testing-manifests/sriovdp-sa.yaml
W0714 19:36:36.927]     
W0714 19:36:36.927] 
W0714 19:36:36.927]     _output/local/go/src/k8s.io/kubernetes/test/e2e_node/device_plugin_test.go:86 
```

fromanirh commented:
> Looks like this is now failing for a new, exciting reason? https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103297/pull-kubernetes-node-kubelet-serial/1415379089107193856/#1:build-log.txt%3A1887
> 
> ```
> W0714 19:36:36.926]     Jul 14 19:29:12.189: Test file "test/e2e/testing-manifests/sample-device-plugin.yaml" was not found.
>       
> W0714 19:36:36.927]     The following files are embedded into the test executable:
> W0714 19:36:36.927]     	test/e2e_node/testing-manifests/sriovdp-cm.yaml
> W0714 19:36:36.927]     	test/e2e_node/testing-manifests/sriovdp-ds.yaml
> W0714 19:36:36.927]     	test/e2e_node/testing-manifests/sriovdp-sa.yaml
> W0714 19:36:36.927]     
> W0714 19:36:36.927] 
> W0714 19:36:36.927]     _output/local/go/src/k8s.io/kubernetes/test/e2e_node/device_plugin_test.go:86 
> ```

Nice, something changed in between. But I'm familiar(ish) with this code, so I'll have a look ASAP

fromanirh commented:
nothing obvious after initial investigation. The relevant manifest is still present even after a rebase against latest master.
Maybe a glitch? let's retry.

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial
the failures are mostly about getting up to date metric at the end of each spec.

ehashman commented:
Will retry after #103865 merges

fromanirh commented:
> Will retry after #103865 merges

Yes, this is the right fix. I managed to reproduce but didn't manage to complete the PR. We could need more fixes in this area (#103865 is the minimal fix), so please don't merge just yet this PR if possible, I'll be updating it ASAP.

dims commented:
/test pull-kubernetes-node-kubelet-serial

fromanirh commented:
/retest
/test pull-kubernetes-node-kubelet-serial

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial

fromanirh commented:
did a retest to check how flaky is this fix. It seems we can improve a bit more.

fromanirh commented:
the ugly sleep is now gone, but not there yet. I'm observing some races/suspect behaviour with kubelet restarts.

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial

ehashman commented:
/unassign

ffromani committed:
e2e: node: split device plugin test

Split the device plugin test, isolating the disruptive and potentially
flaky portion which requires a kubelet restart, which is fragile
and thus lead to flakes at the moment of writing.

This allows us to keep running the portion of the test which is
safe and ha good record trail (almost no flakes recorded).
The refactoring needed to split the test also allows us to
improve the test reliability bringing in more modern constructs
used in the e2e tests lately, and to speed up the cleanup a bit.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/hold
trying a more aggressive refactoring

fromanirh commented:
/hold cancel

cynepco3hahue commented:
/retest

fromanirh commented:
Note none of the failures are related to the test I'm changing here

AlexeyPerevalov commented:
pull-kubernetes-node-kubelet-serial failed due to there is no nvidia.com/gpu in allocatable list after device plugin started, the skip condition in this case is
find /sys/devices/pci* -type f | grep vendor | xargs cat | grep 0x10de
but it's not the way how https://github.com/GoogleCloudPlatform/container-engine-accelerators.git works, so test wasn't skipped, but probably on that host it should. I think test should be improved
pull-kubernetes-node-kubelet-serial-containerd - failed due to there is no schedulable nodes in cluster in system_node_critical_test.go, but this test tries to rich DiskPressure, it should get into node not ready state, probably failure occurs due to:
I0824 17:47:23.054] Aug 24 16:54:10.909: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20210824-f18d4390 is false instead of true. Reason: KubeletNotReady, message: **PLEG** is not healthy: pleg has yet to be successful. As I heart it's well know race.

cynepco3hahue commented:
@fromanirh Do we have a lane that runs `ginkgo.It("[Flaky][Disruptive] Verifies the Kubelet device plugin functionality after kubelet restart."` test case?

fromanirh commented:
a better fix landed meanwhile
